src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -50574,8 +50408,916 @@ class C {}
+
+            var a = comp.GetTypeByMetadataName("A`1");
+            Assert.Equal("A<T> where T : class?", a.ToDisplayString(SymbolDisplayFormat.TestFormatWithConstraints.WithCompilerInternalOptions(SymbolDisplayCompilerInternalOptions.IncludeNonNullableTypeModifier)));
+
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NonNullTypesContext_19()
+        {
+            var source =
+@"
+class A<T> where T : class
+#nullable disable
+    ?
+#nullable enable
+{
+}
+";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+
+            var a = comp.GetTypeByMetadataName("A`1");
+            Assert.Equal("A<T> where T : class?", a.ToDisplayString(SymbolDisplayFormat.TestFormatWithConstraints.WithCompilerInternalOptions(SymbolDisplayCompilerInternalOptions.IncludeNonNullableTypeModifier)));
+
+            comp.VerifyDiagnostics(
+                // (4,5): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' context.
+                //     ?
+                Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(4, 5)
+                );
+        }
+
+        [Fact]
+        public void NonNullTypesContext_20()
+        {
+            var source =
+@"
+class A<T> where T :
+#nullable enable
+    unmanaged
+#nullable disable
+{
+}
+
+class unmanaged {}
+";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesFalse());
+
+            var a = comp.GetTypeByMetadataName("A`1");
+            Assert.Equal("A<T> where T : unmanaged!", a.ToDisplayString(SymbolDisplayFormat.TestFormatWithConstraints.WithCompilerInternalOptions(SymbolDisplayCompilerInternalOptions.IncludeNonNullableTypeModifier)));
+        }
+
+        [Fact]
+        public void NonNullTypesContext_21()
+        {
+            var source =
+@"
+class A<T> where T :
+#nullable disable
+    unmanaged
+#nullable enable
+{
+}
+
+class unmanaged {}
+";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+
+            var a = comp.GetTypeByMetadataName("A`1");
+            Assert.Equal("A<T> where T : unmanaged", a.ToDisplayString(SymbolDisplayFormat.TestFormatWithConstraints.WithCompilerInternalOptions(SymbolDisplayCompilerInternalOptions.IncludeNonNullableTypeModifier)));
+        }
+
+        [Fact]
+        public void NonNullTypesContext_22()
+        {
+            var source =
+@"
+#pragma warning disable CS0169
+
+class A<T>
+{
+    void Tests()
+    {
+        string b;
+    }
+}
+";
+            TypeSyntax type = SyntaxFactory.ParseTypeName(
+@"
+A<string>
+"
+            );
+
+            AssertNonNullTypesContext_22(source, WithNonNullTypesTrue(), type, "A<System.String!>");
+        }
+
+        private static void AssertNonNullTypesContext_22(string source, CSharpCompilationOptions options, TypeSyntax type, string expected)
```

Comment:
>AssertNonNullTypesContext_22 [](start = 28, length = 28)

Since this method is re-used, consider giving it a meaningful name. Perhaps `VerifySpeculativeTypeInfo`.